### PR TITLE
e2e: add a test case for rbd-nbd mounter

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does #

Validate the basic working of rbd-nbd

**Test cases:**
1. Create a PV with rbd-nbd backend and start the application pod using it
2. After the application pod is started, restart the node plugin and expect the IO to fail as the rbd-nbd process is killed
3. After restarting the node plugin, restart the rbd-nbd process by reattaching/re-mapping the device connection and expect the IO from the application pod to continue

### Dependencies ###
- [x] Requires `ceph/ceph:pacific` which is not yet released
- [x] Depends on `nbd.ko` in minikube, see https://github.com/kubernetes/minikube/pull/10217

Updates: #667

### CODE REVIEW

**Comment by pkalever** [comment]
Depends on: https://github.com/kubernetes/minikube/pull/10217

**Comment by nixpanic** [comment]
You may want to have a look at #1840 for an idea how to test it in the CI before the minikube PR is merged+released.

**Comment by pkalever** [comment]
@nixpanic I hope the CI picks the modifications to minikube iso URL as part of this PR testing, is that right?
Or CI will only pick once the changes are merged?

**Comment by nixpanic** [comment]
> @nixpanic I hope the CI picks the modifications to minikube iso URL as part of this PR testing, is that right?
> Or CI will only pick once the changes are merged?

The CI uses `scripts/minikube.sh` from the PR that modifies it, so `--iso-url` will be included in this testing.

**Comment by nixpanic** [comment]
Oh, note that #1831 contains a modification to the `EXTRA_CONFIG` options for minikube, done right. CI jobs will likely fail with the change you made here now.

**Comment by pkalever** [comment]
> Oh, note that #1831 contains a modification to the `EXTRA_CONFIG` options for minikube, done right. CI jobs will likely fail with the change you made here now.

The PR#1831 seems to be in open state now, when it gets merged I shall rebase this, hope that looks like a good plan.
And yeah! while PR#1831 is still in the open state, I don't see a reason why this should fail?

Thanks!

**Comment by pkalever** [comment]
@nixpanic @Madhu-1 @Yuggupta27 any idea why the CI is failing at 'deploy k8s-1.xx and rook' phase?

Thanks!

**Comment by Madhu-1** [comment]
Looks to be a PSP issue https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/567/pipeline

**Comment by pkalever** [comment]
> Looks to be a PSP issue https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/567/pipeline

Thanks @Madhu-1, now I understand why @nixpanic was warning in the comments above.
At least until https://github.com/ceph/ceph-csi/pull/1831/commits/73d9428760ac2cdcda7979bc871e8e90639be23a is merged, I should copy them to my PR, just to comfort CI.

**Comment by pkalever** [comment]
@nixpanic @Madhu-1 @Yuggupta27 Yay! and the CI finally made it, thanks guys!

**Comment by nixpanic** [comment]
With #1811 the `minikube start` commands in `scripts/minikube.sh` have been adapted. There is an additional `--cni` option now. You'll need to manually rebase this.

File: e2e/pod.go [hunk 1]
```diff
@@ -88,21 +88,43 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
...
+func getCommandInPodOpts(f *framework.Framework, c, ns string, cn string, opt *metav1.ListOptions) (framework.ExecOptions, error) {
+	e2elog.Logf("In getCommandInPodOpts: %v", cn)
+	pn := ""
 	cmd := []string{"/bin/sh", "-c", c}
 	podList, err := f.PodClientNS(ns).List(context.TODO(), *opt)
 	framework.ExpectNoError(err)
 	if len(podList.Items) == 0 {
 		return framework.ExecOptions{}, errors.New("podlist is empty")
 	}
+	found := false
```
Comment (humblec) [inline e2e/pod.go]: find Container Name in the pod can be wrapped into a new function .

File: e2e/pod.go [hunk 2]
```diff
@@ -111,7 +133,19 @@ func getCommandInPodOpts(f *framework.Framework, c, ns string, opt *metav1.ListO
...
+	}
+	stdOut, stdErr, err := f.ExecWithOptions(podPot)
+	if stdErr != "" {
+		e2elog.Logf("stdErr occurred: %v", stdErr)
+	}
+	return stdOut, stdErr, err
+}
+
+func execCommandInContainer(f *framework.Framework, c, ns string, cn string, opt *metav1.ListOptions) (string, string, error) {
+	podPot, err := getCommandInPodOpts(f, c, ns, cn, opt)
```
Comment (humblec) [inline e2e/pod.go]: eventhough its not introduced here, Is above (`podPot`) a weird name ? may be we have to rename it as `podOpt` ?

File: e2e/pod.go (hunk 1 above)
Comment (pkalever) [reply e2e/pod.go]: Done now!

File: e2e/pod.go (hunk 2 above)
Comment (pkalever) [reply e2e/pod.go]: addressed this one too. Thanks!

File: e2e/pod.go (hunk 1 above)
Comment (humblec) [reply e2e/pod.go]: Thanks @pkalever :+1:

File: e2e/pod.go (hunk 2 above)
Comment (humblec) [reply e2e/pod.go]: Cool. thanks !

File: e2e/rbd.go [hunk 3]
```diff
@@ -334,6 +335,229 @@ var _ = Describe("RBD", func() {
...
+				// # ps -eo 'cmd' | grep [r]bd-nbd
+				// /usr/bin/rbd-nbd --id cephcsi-rbd-node -m svc-name:6789 --keyfile=/tmp/csi/keys/keyfile attach --device /dev/nbd0 pool-name/image-name --try-netlink --reattach-timeout=180
+				opt1 := metav1.ListOptions{
+					LabelSelector: "app in (ceph-csi-rbd, csi-rbdplugin)",
+				}
+				uname, stdErr, err := execCommandInContainer(f, "uname -a", cephCSINamespace, "csi-rbdplugin", &opt1)
+				if err != nil || stdErr != "" {
+					e2elog.Failf("failed to run uname cmd : %v, stdErr: %v ", err, stdErr)
+				}
+				e2elog.Logf("uname 1: %v", uname)
```
Comment (humblec) [inline e2e/rbd.go]: @pkalever instead of `1,2,..etc` may be we can log other marker?

Comment (pkalever) [reply e2e/rbd.go]: Humble, right now, the helm chart e2e tests are failing for rbd-nbd, so I wanted to debug, hence added some extra debug logs. These logs will not be part of the original patch and I shall remove them once everything runs smooth.

Thanks for keeping an eye on this PR!

File: e2e/rbd.go [hunk 4]
```diff
@@ -427,6 +428,136 @@ var _ = Describe("RBD", func() {
...
+				app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvc.Name
+				err = createPVCAndApp("", f, pvc, app, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to create PVC and application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 1)
+
+				// Get the label name of daemonset dynamically, as the label name is not same for helm and non-helm deploys
+				ds, err := f.ClientSet.AppsV1().DaemonSets(cephCSINamespace).Get(context.TODO(), rbdDaemonsetName, metav1.GetOptions{})
```
Comment (pkalever) [inline e2e/rbd.go]: @nixpanic @Madhu-1 JFYI - this way I'm now able to get the label dynamically.

Comment (Madhu-1) [reply e2e/rbd.go]: Thanks @pkalever

Comment (Madhu-1) [reply e2e/rbd.go]: Use this label selector to delete the daemonset pods also?

Comment (pkalever) [reply e2e/rbd.go]: New line number 509?
I'm doing the same right?
'm I missing anything?

Comment (Madhu-1) [reply e2e/rbd.go]: i was referring to https://github.com/ceph/ceph-csi/pull/1839/files#diff-bfe7fe7ada8a43a75c21ddfd9535e10f37d6e55e9a3006ce5a1872f73f3ae198R399-R402

Comment (pkalever) [reply e2e/rbd.go]: Hmm that was part of a different test scenario. But it's a good idea to keep this logic in the utils and reuse it where ever needed. I will fix it, thanks!

Comment (pkalever) [reply e2e/rbd.go]: @Madhu-1 done now. Thanks!

**Comment by pkalever** [comment]
@nixpanic @Madhu-1 please remove the DNM label, and also help review and merge this PR.
The checklist is now ready.

Thanks!

File: build.env [hunk 5]
```diff
@@ -43,7 +43,7 @@ CHANGE_MINIKUBE_NONE_USER=true
 # Rook options
 ROOK_VERSION=v1.4.9
 # Provide ceph image path
-ROOK_CEPH_CLUSTER_IMAGE=docker.io/ceph/ceph:v14.2.12
+ROOK_CEPH_CLUSTER_IMAGE=docker.io/ceph/ceph:v16
```
Comment (nixpanic) [inline build.env]: duplicate commit of #2038 ?

**Review by nixpanic — COMMENTED** [review]
it seems golang-ci fails, you can run this locally with `make containerized-test` to verify your fix(es)

File: build.env (hunk 5 above)
Comment (pkalever) [reply build.env]: Yes and this is intentional for satisfying the dependency for this PR at CI. I will remove it once the other PR is merged.

**Comment by pkalever** [comment]
> it seems golang-ci fails, you can run this locally with `make containerized-test` to verify your fix(es)

@nixpanic Is there a way to ignore these errors, the errors don't look like a great deal to me, I think we can safely ignore them for now.

File: e2e/pod.go [hunk 6]
```diff
@@ -105,17 +105,49 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
 	return nil
 }
 
-func getCommandInPodOpts(f *framework.Framework, c, ns string, opt *metav1.ListOptions) (framework.ExecOptions, error) {
+func findContainerName(f *framework.Framework, c, ns, cn string, opt *metav1.ListOptions) (string, string, error) {
+	pn := ""
+	podList, err := f.PodClientNS(ns).List(context.TODO(), *opt)
+	framework.ExpectNoError(err)
```
Comment (Madhu-1) [inline e2e/pod.go]: return error?

File: e2e/pod.go [hunk 7]
```diff
@@ -105,17 +105,49 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
 	return nil
 }
 
-func getCommandInPodOpts(f *framework.Framework, c, ns string, opt *metav1.ListOptions) (framework.ExecOptions, error) {
+func findContainerName(f *framework.Framework, c, ns, cn string, opt *metav1.ListOptions) (string, string, error) {
```
Comment (Madhu-1) [inline e2e/pod.go]: the function name should be changed you are finding both pod name and container name here?

File: e2e/pod.go [hunk 8]
```diff
@@ -105,17 +105,49 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
...
+	pn := ""
+	podList, err := f.PodClientNS(ns).List(context.TODO(), *opt)
+	framework.ExpectNoError(err)
+	if len(podList.Items) == 0 {
+		return "", "", errors.New("podlist is empty")
+	}
+	found := false
+	if cn != "" {
+		for i := range podList.Items {
```
Comment (Madhu-1) [inline e2e/pod.go:119]: found is only using if the `cn` is not empty can we move found:=false inside if check?

File: e2e/pod.go [hunk 9]
```diff
@@ -105,17 +105,49 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
...
+			for j := range podList.Items[i].Spec.Containers {
+				if podList.Items[i].Spec.Containers[j].Name == cn {
+					pn = podList.Items[i].Name
+					found = true
+					break
```
Comment (Madhu-1) [inline e2e/pod.go]: if found  return from here? so that we can avoid the new variable `found` and also an if check?

File: e2e/rbd.go [hunk 10]
```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
 				}
 			})
 
+			By("create a PVC and bind it to an app using rbd-nbd mounter", func() {
```
Comment (Madhu-1) [inline e2e/rbd.go:353]: can you please enable this E2E also https://github.com/ceph/ceph-csi/blob/devel/e2e/rbd.go#L485-L496?

File: e2e/rbd.go [hunk 11]
```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
...
+				}
+				pvc, err := loadPVC(pvcPath)
+				if err != nil {
+					e2elog.Failf("failed to load PVC with error %v", err)
+				}
+				pvc.Namespace = f.UniqueName
+
+				app, err := loadApp(appPath)
+				if err != nil {
+					e2elog.Failf("failed to  load application with error %v", err)
```
Comment (Madhu-1) [inline e2e/rbd.go]:
```suggestion
e2elog.Failf("failed to load application with error %v", err)
```

File: e2e/rbd.go [hunk 12]
```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
...
+					e2elog.Failf("timeout waiting for daemonset pods with error %v", err)
+				}
+
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("app=%s", app.Name),
+				}
+				// Write something to mountpoint and expect a failure
+				filePath := app.Spec.Containers[0].VolumeMounts[0].MountPath + "/test"
+				_, stdErr := execCommandInPodAndAllowFail(f, fmt.Sprintf("echo 'Hello World' > %s", filePath), app.Namespace, &opt)
+				IOErr := fmt.Sprintf("cannot create %s: Input/output error", filePath)
```
Comment (Madhu-1) [inline e2e/rbd.go:441]: Add a FIXME to fix this one as this wont be true when we implement logic to handle daemonset pod restart?

File: e2e/rbd.go [hunk 13]
```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
...
+				} else {
+					e2elog.Logf("failed IO as expected: %v", stdErr)
+				}
+
+				err = deletePVCAndApp("", f, pvc, app)
+				if err != nil {
+					e2elog.Failf("failed to delete PVC and application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0)
```
Comment (Madhu-1) [inline e2e/rbd.go:453]: delete the storageclass here as this is specific to `nbr` adding a test for `krbd` in between might get failed because of this one? and recreate it in next By if you want?

File: e2e/rbd.go [hunk 14]
```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
...
+			})
+
+			By("restart rbd-nbd process on nodeplugin and continue IO after nodeplugin restart", func() {
```
Comment (Madhu-1) [inline e2e/rbd.go:464]: Just a question Is this whole test case required? as we are doing some hacks to remount the rbd image which is not the case in the real world and this is expected to handled automatically by cephcsi or other sidecars in daemonset pod?

File: e2e/pod.go (hunk 7 above)
Comment (pkalever) [reply e2e/pod.go]: May be findPodAndContainerName() ?

File: e2e/pod.go (hunk 6 above)
Comment (pkalever) [reply e2e/pod.go]: This is an assertion, we are returning at line 113

File: e2e/pod.go (hunk 9 above)
Comment (pkalever) [reply e2e/pod.go]: Lol, you are right. I think I was called on a Mars mission when I was writing this piece of code.

File: e2e/rbd.go (hunk 13 above)
Comment (pkalever) [reply e2e/rbd.go:453]: sure

File: e2e/rbd.go (hunk 14 above)
Comment (pkalever) [reply e2e/rbd.go:464]: The subject of the tests remains the same after the automation as well, just that we might remove the lines which are trying to manually bring the process online and override them with the steps to bring the process back to online in the background.

In anyway having this in the CI till then will help us understand the regressions and true failures within rbd-nbd.

File: e2e/rbd.go (hunk 13 above)
Comment (pkalever) [reply e2e/rbd.go:453]: done

File: e2e/rbd.go (hunk 12 above)
Comment (pkalever) [reply e2e/rbd.go:441]: sure

File: e2e/rbd.go (hunk 10 above)
Comment (pkalever) [reply e2e/rbd.go:353]: Cool

**Review by pkalever — COMMENTED** [review]
Thanks @Madhu-1

File: e2e/pod.go (hunk 6 above)
Comment (Madhu-1) [reply e2e/pod.go]: what I meant was as we are returning an error from this function, let's return an error from here and the caller handles it. let's not do both assertion and error returning from the same functions

File: e2e/pod.go (hunk 7 above)
Comment (Madhu-1) [reply e2e/pod.go]: Sounds good.

File: e2e/rbd.go (hunk 14 above)
Comment (Madhu-1) [reply e2e/rbd.go:464]: currently this PR adds two tests  `perform IO on rbd-nbd volume after nodeplugin restart and expect a failure` and `restart rbd-nbd process on nodeplugin and continue IO after nodeplugin restart`  but when we bring the automatic handling of the restart, one test will be removed and other will be modified to check we are able to read/write the data? IMO having single test seems fine for me. Lets hear from others

Comment (pkalever) [reply e2e/rbd.go:464]: These are the POC level tests, for someone who is new to rbd-nbd mounter can understand the problem we are dealing here very clearly with the help of these test cases.

In the future when the real story is working we can clear these bits and makeup more useful test cases.

Till then the intention of these tests is to catch some regressions or failures at rbd-nbd only.

File: e2e/pod.go [hunk 15]
```diff
@@ -105,17 +105,40 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
...
+					pn = podList.Items[i].Name
+					return pn, cn, nil
+				}
+			}
+		}
+		return "", "", errors.New("container name not found")
+	}
+	pn = podList.Items[0].Name
+	cn = podList.Items[0].Spec.Containers[0].Name
+	return pn, cn, nil
```
Comment (Madhu-1) [inline e2e/pod.go]: just return podList.Items[0].Name and podList.Items[0].Spec.Containers[0].Name directly here and above

File: e2e/pod.go (hunk 6 above)
Comment (Madhu-1) [reply e2e/pod.go]: @pkalever any comment on this one?

**Review by Madhu-1 — COMMENTED** [review]
don't have a major comment except for https://github.com/ceph/ceph-csi/pull/1839/files#r624912403 if others are okay we can get it in.

File: e2e/pod.go (hunk 15 above)
Comment (pkalever) [reply e2e/pod.go]: Sure.

File: e2e/pod.go (hunk 6 above)
Comment (pkalever) [reply e2e/pod.go]: Ah! Got you, I will remove this line, thanks for the clarification.

File: e2e/pod.go [hunk 16]
```diff
@@ -305,7 +305,7 @@ func deletePod(name, ns string, c kubernetes.Interface, t int) error {
 	})
 }
 
-func deletePodWithLabel(label, ns string, skipNotFound bool) error {
+func deletePodWithLabel(label, ns string, skipNotFound bool) error { //nolint:unparam
```
Comment (pkalever) [inline e2e/pod.go]: @nixpanic this should fix the golint warnings!

**Comment by pkalever** [comment]
@Madhu-1 @nixpanic As we don't see any objections, could you please merge this PR now?

Thanks!

**Comment by humblec** [comment]
@pkalever one quick check: Have we tested this scenario when the actual node reboot or move to `NOT READY` state and come back after some time ?

**Comment by pkalever** [comment]
> @pkalever one quick check: Have we tested this scenario when the actual node reboot or move to `NOT READY` state and come back after some time ?

When a node is rebooted, the application pod might be migrated to a new node or might be scheduled on the same node, in any case, a new NodeStageVolume call is expected to take care of mapping the device.

These tests are mainly focusing on the restart of the node plugin, where there is no NodeStageVolume call for now. We are working on implementing the missing gaps in parallel, where a new init container or a sidecar is supposed to get the list of volume attachments on a node (upon the restart of the node plugin) and make new NodeStageVolume grpc Calls per device for reattaching (remapping) rbd-nbd processes/devices.

Thanks!

**Comment by humblec** [comment]
> > @pkalever one quick check: Have we tested this scenario when the actual node reboot or move to `NOT READY` state and come back after some time ?
> 
> When a node is rebooted, the application pod might be migrated to a new node or might be scheduled on the same node, in any case, a new NodeStageVolume call is expected to take care of mapping the device.

There are some ( corner ) cases, like network error, kubelet goes wrong, node completely goes out..etc which are bit different from volume pov. We have also seen issues popping up with some race conditions from CO to detect and issues RPC calls in between..etc.  The reason for asking node `not ready` scenario testing  was to understand/make sure we are not landing into issues easily when we remount the shares with these options.  So I think its good to check once if we have a setup :)

**Comment by pkalever** [comment]
> There are some ( corner ) cases, like network error, kubelet goes wrong, node completely goes out..etc which are bit different from volume pov. We have also seen issues popping up with some race conditions from CO to detect and issues RPC calls in between..etc. The reason for asking node `not ready` scenario testing was to understand/make sure we are not landing into issues easily when we remount the shares with these options. So I think its good to check once if we have a setup :)

Thanks Humble, Got you. These test cases are less focused on the corner case. But I will make sure to test these corner cases with the logic we are implementing to address node plugin restart automatically.

**Comment by pkalever** [comment]
@Madhu-1 @nixpanic can you please help on taking this PR to merge?
I would like to see them running in the CI which will help us improve the confidence on rbd-nbd.

Thanks!

**Review by Madhu-1 — APPROVED** [review]

File: e2e/pod.go [hunk 17]
```diff
@@ -204,7 +204,7 @@ func execCommandInPod(f *framework.Framework, c, ns string, opt *metav1.ListOpti
 	return stdOut, stdErr, err
 }
 
-func execCommandInContainer(f *framework.Framework, c, ns, cn string, opt *metav1.ListOptions) (string, string, error) {
+func execCommandInContainer(f *framework.Framework, c, ns, cn string, opt *metav1.ListOptions) (string, string, error) { //nolint:unparam,lll // cn can be used with different inputs later
```
Comment (nixpanic) [inline e2e/pod.go:207]: We usually put these comments above the line that is affected, not after the line (this line becomes *really* long now).

**Review by nixpanic — APPROVED** [review]
It is good to have this tested regularly, so let's get it in.

In case a manual rebase is needed, please correct the formatting of the `// nolint:` comments.

File: e2e/pod.go (hunk 17 above)
Comment (pkalever) [reply e2e/pod.go:207]: I thought putting it above the line will affect the whole function block not just that line in this case?

**Comment by pkalever** [comment]
> In case a manual rebase is needed, please correct the formatting of the `// nolint:` comments.

The recommendation was not to have space in between the slash and `nolint`

File: e2e/pod.go (hunk 17 above)
Comment (nixpanic) [reply e2e/pod.go:207]: Probably, but that is not really an issue.

In general, all `nolint` issues need to be addressed at one point. That means, either the argument is not needed and can be dropped, or the function gets used with other arguments somewhere.

**Comment by Madhu-1** [comment]
@Mergifyio rebase
